types/crawlers/puppeteer_crawler.d.ts

```diff
@@ -0,0 +1,505 @@
+ * it is useful for crawling of websites that require to execute JavaScript.
+ * If the target website doesn't need JavaScript, consider using {@link CheerioCrawler},
+ * which downloads the pages using raw HTTP requests and is about 10x faster.
+ *
+ * The source URLs are represented using {@link Request} objects that are fed from
+ * {@link RequestList} or {@link RequestQueue} instances provided by the [`requestList`](#new_PuppeteerCrawler_new)
+ * or [`requestQueue`](#new_PuppeteerCrawler_new) constructor options, respectively.
+ *
+ * If both [`requestList`](#new_PuppeteerCrawler_new) and [`requestQueue`](#new_PuppeteerCrawler_new) are used,
+ * the instance first processes URLs from the {@link RequestList} and automatically enqueues all of them
+ * to {@link RequestQueue} before it starts their processing. This ensures that a single URL is not crawled multiple times.
+ *
+ * The crawler finishes when there are no more {@link Request} objects to crawl.
+ *
+ * `PuppeteerCrawler` opens a new Chrome page (i.e. tab) for each {@link Request} object to crawl
+ * and then calls the function provided by user as the [`handlePageFunction()`](#new_PuppeteerCrawler_new) option.
+ *
+ * New pages are only opened when there is enough free CPU and memory available,
+ * using the functionality provided by the {@link AutoscaledPool} class.
+ * All {@link AutoscaledPool} configuration options can be passed to the `autoscaledPoolOptions`
+ * parameter of the `PuppeteerCrawler` constructor. For user convenience, the `minConcurrency` and `maxConcurrency`
+ * {@link AutoscaledPool} options are available directly in the `PuppeteerCrawler` constructor.
+ *
+ * Note that the pool of Puppeteer instances is internally managed by
+ * the {@link PuppeteerPool} class. Many constructor options
+ * such as `maxOpenPagesPerInstance` or `launchPuppeteerFunction` are passed directly
+ * to {@link PuppeteerPool} constructor.
+ *
+ * **Example usage:**
+ *
+ * ```javascript
+ * const crawler = new Apify.PuppeteerCrawler({
+ *     requestList,
+ *     handlePageFunction: async ({ page, request }) => {
+ *         // This function is called to extract data from a single web page
+ *         // 'page' is an instance of Puppeteer.Page with page.goto(request.url) already called
+ *         // 'request' is an instance of Request class with information about the page to load
+ *         await Apify.pushData({
+ *             title: await page.title(),
+ *             url: request.url,
+ *             succeeded: true,
+ *         })
+ *     },
+ *     handleFailedRequestFunction: async ({ request }) => {
+ *         // This function is called when the crawling of a request failed too many times
+ *         await Apify.pushData({
+ *             url: request.url,
+ *             succeeded: false,
+ *             errors: request.errorMessages,
+ *         })
+ *     },
+ * });
+ *
+ * await crawler.run();
+ * ```
+ */
+declare class PuppeteerCrawler {
```

Comment:
could this take a generic T, and provide to the `RequestQueue<T>` (so it can infer from options), so user data inside the `handlePageFunction` / `PuppeteerHandlePage<T>` is automatically typed? it would apply to the other crawlers as well

Reply:
This is non-critical, so let's postpone this for later.